### PR TITLE
[Feat] Support Torch Symm Mem AllReduce

### PR DESCRIPTION
**Pull request by yuan-luo**
## Motivation

This PR is to support PyTorch Symm Mem AllReduce which support large message package size.

Delivers performance gains for scenarios with msg_size > 512KB.
Currently, MSCCLPP only supports a maximum message size of 1MB; messages larger than 1MB will error out.
Symm Mem AllReduce supports messages ≥ 2MB.
This PR lays the groundwork for long-context support and enables large-message AllReduce (AR).

Torch Symm Mem AllReduce is based on NVLS, not using SM.
On Blackwell, NVLS is always deterministic.
On Hopper, NVLS is deterministic with cuda 12.8+ or updated cuda 12.4 

image

image

```
SYMM_MEM:
| msg_size   |   torch eager time |   symm mem eager time |   symm mem graph time |   pynccl graph time |
|------------|--------------------|-----------------------|-----------------------|---------------------|
| 2.0 KiB    |            48.6752 |               42.6112 |               8.95296 |             16.4637 |
| 4.0 KiB    |            26.304  |               27.6096 |               8.99104 |             18.583  |
| 8.0 KiB    |            21.9744 |               27.3632 |               9.21984 |             18.6733 |
| 16.0 KiB   |            20.3968 |               27.1584 |               9.3568  |             19.1792 |
| 32.0 KiB   |            20.2848 |               34.1888 |               9.40416 |             20.8093 |
| 64.0 KiB   |            20.5088 |               26.5888 |               9.73728 |             20.968  |
| 128.0 KiB  |           110.179  |               26.8448 |              10.0224  |             21.5664 |
| 256.0 KiB  |            20.9184 |               29.7504 |              10.6275  |             21.7779 |
| 512.0 KiB  |            23.5904 |               27.5392 |              12.2198  |             21.9856 |
| 1.0 MiB    |            28.1664 |               29.1744 |              14.2813  |             27.6038 |
|------------|--------------------|-----------------------|-----------------------|---------------------|
| 2.0 MiB    |            39.2576 |               29.1744 |              19.0234  |             39.5312 |  <<<<<< MSCCLPP doesn't support from this size on
| 4.0 MiB    |            52.5408 |               33.4528 |              28.1082  |             52.8202 |
| 8.0 MiB    |            78.064  |               74.0832 |              46.6918  |             86.2893 |
| 16.0 MiB   |           118.87   |               93.2256 |              88.1763  |            119.501  |

MSCCLPP:
| msg_size   |   torch eager time |   msccl eager time |   msccl graph time |   pynccl graph time |
|------------|--------------------|--------------------|--------------------|---------------------|
| 2.0 KiB    |            67.6608 |            19.7952 |            7.17696 |             18.5763 |
| 4.0 KiB    |            22.7936 |            19.248  |            6.94688 |             18.6624 |
| 8.0 KiB    |            20.5888 |            19.4688 |            7.32512 |             19.0259 |
| 16.0 KiB   |            21.4752 |            20.9152 |            7.58144 |             19.3904 |
| 32.0 KiB   |            21.4208 |            21.1424 |            7.55168 |             20.9411 |
| 64.0 KiB   |            21.84   |            19.8304 |            8.00928 |             21.1987 |
| 128.0 KiB  |            20.9728 |            20.72   |            8.91872 |             21.479  |
| 256.0 KiB  |            24.2208 |            20.2752 |            9.8064  |             21.7882 |
| 512.0 KiB  |            24.4448 |            21.9616 |           11.9722  |             22.0272 |
| 1.0 MiB    |            26.8448 |            22.1696 |           18.2694  |             27.7914 |
```

## Accuracy Tests

gsm8k DeepSeek-R1, accuracy the same as main.
```
PR:

Server Command:
$python3 -m sglang.launch_server --model-path /home/admin/DeepSeek-R1 --host 0.0.0.0 --port 30000 --trust-remote-code --enable-cache-report --quantization fp8 --log-level info --max-running-requests 32 --mem-fraction-static 0.92 --chunked-prefill-size 16384 --context-length 65535 --attention-backend flashinfer --disable-radix-cache --page-size 64 --tp-size 8 --enable-metrics --cuda-graph-max-bs 32 --enable-torch-symm-mem
INFO 10-02 21:34:31 [__init__.py:216] Automatically detected platform cuda.
WARNING:sglang.srt.server_args:
########################################################################
# For contributors and developers:                                    #
# Please move environment variable definitions to sglang.srt.environ  #
# using the following pattern:                                        #
#     SGLANG_XXX = EnvBool(False)                                     #
#                                                                     #
########################################################################

[2025-10-02 21:34:32] server_args=ServerArgs(model_path='/home/admin/DeepSeek-R1', tokenizer_path='/home/admin/DeepSeek-R1', tokenizer_mode='auto', tokenizer_worker_num=1, skip_tokenizer_init=False, load_format='auto', model_loader_extra_config='{}', trust_remote_code=True, context_length=65535, is_embedding=False, enable_multimodal=None, revision=None, model_impl='auto', host='0.0.0.0', port=30000, skip_server_warmup=False, warmups=None, nccl_port=None, dtype='auto', quantization='fp8', quantization_param_path=None, kv_cache_dtype='auto', enable_fp32_lm_head=False, mem_fraction_static=0.92, max_running_requests=32, max_queued_requests=None, max_total_tokens=None, chunked_prefill_size=16384, max_prefill_tokens=16384, schedule_policy='fcfs', enable_priority_scheduling=False, schedule_low_priority_values_first=False, priority_scheduling_preemption_threshold=10, schedule_conservativeness=1.0, page_size=64, hybrid_kvcache_ratio=None, swa_full_tokens_ratio=0.8, disable_hybrid_swa_memory=False, radix_eviction_policy='lru', device='cuda', tp_size=8, pp_size=1, max_micro_batch_size=None, stream_interval=1, stream_output=False, random_seed=563291126, constrained_json_whitespace_pattern=None, watchdog_timeout=300, dist_timeout=None, download_dir=None, base_gpu_id=0, gpu_id_step=1, sleep_on_idle=False, log_level='info', log_level_http=None, log_requests=False, log_requests_level=2, crash_dump_folder=None, show_time_cost=False, enable_metrics=True, enable_metrics_for_all_schedulers=False, tokenizer_metrics_custom_labels_header='x-custom-labels', tokenizer_metrics_allowed_custom_labels=None, bucket_time_to_first_token=None, bucket_inter_token_latency=None, bucket_e2e_request_latency=None, collect_tokens_histogram=False, prompt_tokens_buckets=None, generation_tokens_buckets=None, decode_log_interval=40, enable_request_time_stats_logging=False, kv_events_config=None, gc_warning_threshold_secs=0.0, enable_trace=False, oltp_traces_endpoint='localhost:4317', api_key=None, served_model_name='/home/admin/DeepSeek-R1', weight_version='default', chat_template=None, completion_template=None, file_storage_path='sglang_storage', enable_cache_report=True, reasoning_parser=None, tool_call_parser=None, tool_server=None, dp_size=1, load_balance_method='round_robin', load_watch_interval=0.1, prefill_round_robin_balance=False, dist_init_addr=None, nnodes=1, node_rank=0, json_model_override_args='{}', preferred_sampling_params=None, enable_lora=None, max_lora_rank=None, lora_target_modules=None, lora_paths=None, max_loaded_loras=None, max_loras_per_batch=8, lora_backend='triton', max_lora_chunk_size=16, attention_backend='flashinfer', decode_attention_backend=None, prefill_attention_backend=None, sampling_backend='flashinfer', grammar_backend='xgrammar', mm_attention_backend=None, speculative_algorithm=None, speculative_draft_model_path=None, speculative_draft_model_revision=None, speculative_num_steps=None, speculative_eagle_topk=None, speculative_num_draft_tokens=None, speculative_accept_threshold_single=1.0, speculative_accept_threshold_acc=1.0, speculative_token_map=None, speculative_attention_mode='prefill', speculative_ngram_min_match_window_size=1, speculative_ngram_max_match_window_size=12, speculative_ngram_min_bfs_breadth=1, speculative_ngram_max_bfs_breadth=10, speculative_ngram_match_type='BFS', speculative_ngram_branch_length=18, speculative_ngram_capacity=10000000, ep_size=1, moe_a2a_backend='none', moe_runner_backend='auto', flashinfer_mxfp4_moe_precision='default', enable_flashinfer_allreduce_fusion=False, deepep_mode='auto', ep_num_redundant_experts=0, ep_dispatch_algorithm='static', init_expert_location='trivial', enable_eplb=False, eplb_algorithm='auto', eplb_rebalance_num_iterations=1000, eplb_rebalance_layers_per_chunk=None, eplb_min_rebalancing_utilization_threshold=1.0, expert_distribution_recorder_mode=None, expert_distribution_recorder_buffer_size=1000, enable_expert_distribution_metrics=False, deepep_config=None, moe_dense_tp_size=None, max_mamba_cache_size=None, mamba_ssm_dtype='float32', enable_hierarchical_cache=False, hicache_ratio=2.0, hicache_size=0, hicache_write_policy='write_through', hicache_io_backend='kernel', hicache_mem_layout='layer_first', hicache_storage_backend=None, hicache_storage_prefetch_policy='best_effort', hicache_storage_backend_extra_config=None, enable_lmcache=False, enable_double_sparsity=False, ds_channel_config_path=None, ds_heavy_channel_num=32, ds_heavy_token_num=256, ds_heavy_channel_type='qk', ds_sparse_decode_threshold=4096, cpu_offload_gb=0, offload_group_size=-1, offload_num_in_group=1, offload_prefetch_step=1, offload_mode='cpu', disable_radix_cache=True, cuda_graph_max_bs=32, cuda_graph_bs=[1, 2, 4, 8, 12, 16, 24, 32], disable_cuda_graph=False, disable_cuda_graph_padding=False, enable_profile_cuda_graph=False, enable_cudagraph_gc=False, enable_nccl_nvls=False, enable_symm_mem=False, disable_flashinfer_cutlass_moe_fp4_allgather=False, enable_tokenizer_batch_encode=False, disable_outlines_disk_cache=False, disable_custom_all_reduce=False, enable_mscclpp=False, enable_torch_symm_mem=True, disable_overlap_schedule=False, enable_mixed_chunk=False, enable_dp_attention=False, enable_dp_lm_head=False, enable_two_batch_overlap=False, enable_single_batch_overlap=False, tbo_token_distribution_threshold=0.48, enable_torch_compile=False, torch_compile_max_bs=32, torchao_config='', enable_nan_detection=False, enable_p2p_check=False, triton_attention_reduce_in_fp32=False, triton_attention_num_kv_splits=8, triton_attention_split_tile_size=None, num_continuous_decode_steps=1, delete_ckpt_after_loading=False, enable_memory_saver=False, allow_auto_truncate=False, enable_custom_logit_processor=False, flashinfer_mla_disable_ragged=False, disable_shared_experts_fusion=False, disable_chunked_prefix_cache=False, disable_fast_image_processor=False, keep_mm_feature_on_device=False, enable_return_hidden_states=False, scheduler_recv_interval=1, numa_node=None, enable_deterministic_inference=False, enable_dynamic_batch_tokenizer=False, dynamic_batch_tokenizer_batch_size=32, dynamic_batch_tokenizer_batch_timeout=0.002, debug_tensor_dump_output_folder=None, debug_tensor_dump_input_file=None, debug_tensor_dump_inject=False, debug_tensor_dump_prefill_only=False, disaggregation_mode='null', disaggregation_transfer_backend='mooncake', disaggregation_bootstrap_port=8998, disaggregation_decode_tp=None, disaggregation_decode_dp=None, disaggregation_prefill_pp=1, disaggregation_ib_device=None, disaggregation_decode_enable_offload_kvcache=False, num_reserved_decode_tokens=512, disaggregation_decode_polling_interval=1, custom_weight_loader=[], weight_loader_disable_mmap=False, remote_instance_weight_loader_seed_instance_ip=None, remote_instance_weight_loader_seed_instance_service_port=None, remote_instance_weight_loader_send_weights_group_ports=None, enable_pdmux=False, sm_group_num=3)
`torch_dtype` is deprecated! Use `dtype` instead!
......

Client Side:
$python3 benchmark/gsm8k/bench_sglang.py --num-questions 200 --parallel 128 --num-shots 8 --port 30000 --data-path ./test.jsonl
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [01:01<00:00,  3.25it/s]
Accuracy: 0.980
Invalid: 0.000
Latency: 61.959 s
```

## Benchmarking and Profiling

Benchmark
8k input, 1 output
max-concurrency 16
input throughput speedup 1.4%，
TTFT drop 1.6%：17728.41 vs 18017.47

```
$python3 -m sglang.bench_serving --backend sglang --dataset-name random --dataset-path /mnt/ShareGPT_V3_unfiltered_cleaned_split.json --random-input-len 8192 --random-output-len 1 --max-concurrency 16 --num-prompts 100 --request-rate inf --random-range-ratio 1 
With PR:
$python3 -m sglang.launch_server --model-path /home/admin/DeepSeek-R1 --host 0.0.0.0 --port 30000 --trust-remote-code --enable-cache-report --quantization fp8 --log-level info --max-running-requests 32 --mem-fraction-static 0.92 --chunked-prefill-size 16384 --context-length 65535 --attention-backend flashinfer --disable-radix-cache --page-size 64 --tp-size 8 --enable-metrics --cuda-graph-max-bs 32 --enable-torch-symm-mem
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 16        
Successful requests:                     100       
Benchmark duration (s):                  117.46    
Total input tokens:                      819200    
Total generated tokens:                  100       
Total generated tokens (retokenized):    100       
Request throughput (req/s):              0.85      
Input token throughput (tok/s):          6974.25   
Output token throughput (tok/s):         0.85      
Total token throughput (tok/s):          6975.10   
Concurrency:                             15.09     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   17728.42  
Median E2E Latency (ms):                 18940.88  
---------------Time to First Token----------------
Mean TTFT (ms):                          17728.41  
Median TTFT (ms):                        18940.86  
P99 TTFT (ms):                           21778.77  
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================

Main:
$python3 -m sglang.launch_server --model-path /home/admin/DeepSeek-R1 --host 0.0.0.0 --port 30000 --trust-remote-code --enable-cache-report --quantization fp8 --log-level info --max-running-requests 32 --mem-fraction-static 0.92 --chunked-prefill-size 16384 --context-length 65535 --attention-backend flashinfer --disable-radix-cache --page-size 64 --tp-size 8 --enable-metrics --cuda-graph-max-bs 32
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 16        
Successful requests:                     100       
Benchmark duration (s):                  119.27    
Total input tokens:                      819200    
Total generated tokens:                  100       
Total generated tokens (retokenized):    100       
Request throughput (req/s):              0.84      
Input token throughput (tok/s):          6868.31   
Output token throughput (tok/s):         0.84      
Total token throughput (tok/s):          6869.15   
Concurrency:                             15.11     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   18017.49  
Median E2E Latency (ms):                 19093.72  
---------------Time to First Token----------------
Mean TTFT (ms):                          18017.47  
Median TTFT (ms):                        19093.71  
P99 TTFT (ms):                           21802.45  
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by trevor-m** [comment]
Hi @yuan-luo https://github.com/sgl-project/sglang/pull/8238 enabled NCCL symmetric memory all-reduce with `--enable-symm-mem`. It avoids a copy into the symm memory buffer by using a custom memory allocation pool.
That approach used pynccl because native torch support for symmetric memory wasn't added yet.

**Comment by yuan-luo** [comment]
> Hi @yuan-luo #8238 enabled NCCL symmetric memory all-reduce with `--enable-symm-mem`. It avoids a copy into the symm memory buffer by using a custom memory allocation pool. That approach used pynccl because native torch support for symmetric memory wasn't added yet.

@trevor-m I knew this PR. Both are two different mechnisms. One uses NCCL and the other uses Torch mechanism. I'll add a new feature flag to distinguish them.
Per test result the Torch Symm Mem can provide more robust and high performance AllReduce in large package size. 
Btw, vLLM has set Torch Symmetrix Memory All-Reduce as the default All-Reduce which was custom All-Reduce previously.

**Comment by yuan-luo** [comment]
New server_args _--enable-torch-symm-mem_  E2E test passed.
```
$python3 -m sglang.launch_server --model-path /home/admin/DeepSeek-R1 --host 0.0.0.0 --port 30000 --trust-remote-code --enable-cache-report --quantization fp8 --log-level info --max-running-requests 32 --mem-fraction-static 0.92 --chunked-prefill-size 16384 --context-length 65535 --attention-backend flashinfer --disable-radix-cache --page-size 64 --tp-size 8 --enable-metrics --cuda-graph-max-bs 32 --enable-torch-symm-mem
......

[2025-09-18 17:13:48] INFO:     Started server process [80415]
[2025-09-18 17:13:48] INFO:     Waiting for application startup.
[2025-09-18 17:13:48] INFO:     Application startup complete.
[2025-09-18 17:13:48] INFO:     Uvicorn running on http://0.0.0.0:30000 (Press CTRL+C to quit)
[2025-09-18 17:13:49] INFO:     127.0.0.1:46614 - "GET /get_model_info HTTP/1.1" 200 OK
[2025-09-18 17:13:49 TP0] Prefill batch. #new-seq: 1, #new-token: 64, #cached-token: 0, token usage: 0.00, #running-req: 0, #queue-req: 0, 
[2025-09-18 17:13:49 TP0] Entering DeepGEMM JIT Pre-Compile session. It may take a long time (typically 10-20 mins) if you have not run `sglang.compile_deep_gemm`. It is recommended to run `sglang.compile_deep_gemm` with same args as `sglang.launch_server` for pre-compilation to reduce the overhead if you have not run it before. For example: `python3 -m sglang.compile_deep_gemm --model deepseek-ai/DeepSeek-V3 --tp 8 --trust-remote-code`
[2025-09-18 17:13:49 TP0] Try DeepGEMM JIT Compiling for <GEMM_NT_F8F8BF16> N=4096, K=512, num_groups=1 with all Ms. It only takes a little time (typically 1 sec) if you have run `python3 -m sglang.compile_deep_gemm`. 
DeepGEMM warmup: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 32768/32768 [00:06<00:00, 5397.02it/s]
[2025-09-18 17:13:58] INFO:     127.0.0.1:46618 - "POST /generate HTTP/1.1" 200 OK
[2025-09-18 17:13:58] The server is fired up and ready to roll!
[2025-09-18 17:14:19 TP0] Prefill batch. #new-seq: 1, #new-token: 64, #cached-token: 0, token usage: 0.00, #running-req: 0, #queue-req: 0, 
[2025-09-18 17:14:19 TP0] Decode batch. #running-req: 1, #token: 64, token usage: 0.00, cuda graph: True, gen throughput (token/s): 1.26, #queue-req: 0, 
[2025-09-18 17:14:20 TP0] Decode batch. #running-req: 1, #token: 128, token usage: 0.00, cuda graph: True, gen throughput (token/s): 84.71, #queue-req: 0, 
[2025-09-18 17:14:20 TP0] Decode batch. #running-req: 1, #token: 192, token usage: 0.00, cuda graph: True, gen throughput (token/s): 84.76, #queue-req: 0, 
[2025-09-18 17:14:21 TP0] Decode batch. #running-req: 1, #token: 192, token usage: 0.00, cuda graph: True, gen throughput (token/s): 84.95, #queue-req: 0, 
[2025-09-18 17:14:21 TP0] Decode batch. #running-req: 1, #token: 256, token usage: 0.00, cuda graph: True, gen throughput (token/s): 84.86, #queue-req: 0, 
[2025-09-18 17:14:21] INFO:     127.0.0.1:39634 - "POST /v1/chat/completions HTTP/1.1" 200 OK
```

Client:
```
$cat test_openai.py
import openai
client = openai.Client(base_url="http://127.0.0.1:30000/v1", api_key="EMPTY")
# Chat completion
response = client.chat.completions.create(
    model="default",
    messages=[
        {"role": "system", "content": "You are a helpful AI assistant"},
        {"role": "user", "content": "List 3 countries and their capitals. Tell me how you rank them"},
    ],
    temperature=0,
    max_tokens=200,
)
print(response)

$python test_openai.py 
ChatCompletion(id='b060b6252b784549b14f4268ba7a3c47', choices=[Choice(finish_reason='length', index=0, logprobs=None, message=ChatCompletionMessage(content="Okay, the user wants me to list three countries and their capitals and then rank them. Let me start by picking three countries. Maybe I should choose well-known ones to keep it simple. Let's see, France with Paris, Japan with Tokyo, and Brazil with Brasília. That covers different continents.\n\nNow, how to rank them. The user didn't specify the criteria, so I need to decide on a basis. Population? Economic size? Area? Or maybe cultural significance. Since capitals are involved, maybe the population of the capital cities? Or perhaps the historical importance of the capitals. Alternatively, alphabetical order, but that's not really a ranking. \n\nWait, the user might want a subjective ranking based on some interesting factors. Let me think. If I rank them by the population of the capitals, Tokyo is the largest, then Paris, then Brasília. But maybe the user wants a more engaging reason. For example, cultural influence, tourism, or something like that. \n\n", refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None, reasoning_content=None), matched_stop=None)], created=1758186861, model='default', object='chat.completion', service_tier=None, system_fingerprint=None, usage=CompletionUsage(completion_tokens=200, prompt_tokens=25, total_tokens=225, completion_tokens_details=None, prompt_tokens_details=None, reasoning_tokens=0), metadata={'weight_version': 'default'})
```

File: python/sglang/srt/distributed/device_communicators/symm_mem.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+        if not symm_mem_available:
+            return
+
+        if isinstance(device, int):
+            device = torch.device(f"cuda:{device}")
+        elif isinstance(device, str):
+            device = torch.device(device)
+        torch.cuda.set_device(device)
+        self.dtype = torch.bfloat16
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/distributed/device_communicators/symm_mem.py:73]: symm mem currently prefers to use bfloat16. Keep unchanged.

**Review by Alcanderian — APPROVED** [review]

File: python/sglang/srt/distributed/device_communicators/symm_mem.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    try:
+        import sgl_kernel
+
+        symm_mem_is_available = True
+    except:
+        symm_mem_is_available = False
+
+
+class SymmMemCommunicator:
+    _WORLD_SIZES_MULTIMEM = {
```
Comment (BBuf) [inline python/sglang/srt/distributed/device_communicators/symm_mem.py:51]: Add some commnet is better?

File: python/sglang/srt/distributed/device_communicators/all_reduce_utils.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+SYMM_MEM_ALL_REDUCE_MAX_SIZES = {
+    9: {
+        2: 64 * MiB,  # 64 MB
+        4: 32 * MiB,  # 32 MB
+        6: 64 * MiB,  # 64 MB
+        8: 64 * MiB,  # 64 MB
+    },
+    10: {
+        2: 8 * MiB,  # 8 MB
```
Comment (BBuf) [inline python/sglang/srt/distributed/device_communicators/all_reduce_utils.py]: Why is Blackwell's TP2 configuration (8MB) smaller than Hopper's (64MB)? This seems unreasonable, please verify.

File: test/srt/test_symm_mem_allreduce.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+class TestSymmMemAllReduce(CustomTestCase):
+    @classmethod
+    def setUpClass(cls):
+        random.seed(42)
+        # 1KB to 1MB
+        cls.test_sizes = [512, 4096, 32768, 262144, 524288]
+        cls.world_sizes = [8]
+        TEST_TP16 = int(os.getenv("SGL_SYMM_MEM_TEST_TP16", "0"))
+        if TEST_TP16:
+            cls.world_sizes = [16]
```
[the remark this answers is not in the document]
Comment (BBuf) [reply test/srt/test_symm_mem_allreduce.py]: +1

File: python/sglang/srt/distributed/device_communicators/symm_mem.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+_is_cuda = is_cuda()
+_is_hip = is_hip()
+
+symm_mem_is_available = False
+if _is_hip:
+    symm_mem_is_available = False
+if _is_cuda:
+    try:
+        import sgl_kernel
```
Comment (BBuf) [inline python/sglang/srt/distributed/device_communicators/symm_mem.py]: Why import sgl_kernel succesed means `symm_mem_is_available = True`?

File: python/sglang/srt/server_args.py [hunk 6]
```diff
@@ -2419,6 +2420,11 @@ def add_cli_args(parser: argparse.ArgumentParser):
             action="store_true",
             help="Enable using mscclpp for small messages for all-reduce kernel and fall back to NCCL.",
         )
+        parser.add_argument(
```
Comment (BBuf) [inline python/sglang/srt/server_args.py:2437]: Please add hardware limit and world_size limit.

File: python/sglang/srt/distributed/device_communicators/symm_mem.py (hunk 1 above)
Comment (BBuf) [inline python/sglang/srt/distributed/device_communicators/symm_mem.py:73]: Implementation only supports `bfloat16`, but test code (`test_symm_mem_allreduce.py:189`) tests `float32`, `float16`, and `bfloat16`, this inconsistency will cause test failures?

File: python/sglang/srt/distributed/device_communicators/all_reduce_utils.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
+MiB = 1024 * 1024
+
+SYMM_MEM_ALL_REDUCE_MAX_SIZES = {
+    9: {
+        2: 64 * MiB,  # 64 MB
+        4: 32 * MiB,  # 32 MB
+        6: 64 * MiB,  # 64 MB
+        8: 64 * MiB,  # 64 MB
```
Comment (BBuf) [inline python/sglang/srt/distributed/device_communicators/all_reduce_utils.py:8]: Test code includes TP16 support (`test_symm_mem_allreduce.py:86`), but configuration doesn't?

File: python/sglang/srt/distributed/device_communicators/symm_mem.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+            logger.warning(
+                "SymmMemCommunicator: World size %d not supported, "
+                "communicator is not available.",
+                self.world_size,
+            )
+            return
+        self.max_size = SYMM_MEM_ALL_REDUCE_MAX_SIZES[self.device_capability][
+            self.world_size
+        ]
+        self.buffer = torch_symm_mem.empty(
```
Comment (BBuf) [inline python/sglang/srt/distributed/device_communicators/symm_mem.py:95]: The buffer can be auto released?

File: test/srt/test_symm_mem_allreduce.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
+"""For Now, SymmMem is only supported on TP16 and TP8 case
+
```
Comment (BBuf) [inline test/srt/test_symm_mem_allreduce.py]: The test is dead code? Please add it to ci.

File: test/srt/test_symm_mem_allreduce.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+            )
+        )
+    ray.get(refs)
+
+    ray.shutdown()
+
+
+class TestSymmMemAllReduce(CustomTestCase):
+    @classmethod
+    def setUpClass(cls):
```
Comment (BBuf) [inline test/srt/test_symm_mem_allreduce.py]: Tests don't cover some edge cases:
- Message size exactly equal to `max_size`
- Message size exceeding `max_size` (should fallback to NCCL)
- Unsupported world sizes
- Unsupported dtypes

File: python/sglang/srt/distributed/device_communicators/symm_mem.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+            return
+        self.disabled = False
+
+    def should_symm_mem_allreduce(self, inp: torch.Tensor):
+        if self.disabled:
+            return False
+        if inp.dtype != self.dtype:
+            return False
+        inp_size = inp.numel() * inp.element_size()
+        if inp_size % 4 != 0:
```
[the remark this answers is not in the document]
Comment (BBuf) [reply python/sglang/srt/distributed/device_communicators/symm_mem.py:130]: +1

File: python/sglang/srt/distributed/device_communicators/symm_mem.py (hunk 5 above)
Comment (yuan-luo) [reply python/sglang/srt/distributed/device_communicators/symm_mem.py]: removed sgl_kernel check.

File: python/sglang/srt/server_args.py (hunk 6 above)
Comment (yuan-luo) [reply python/sglang/srt/server_args.py:2437]: Added.

File: python/sglang/srt/distributed/device_communicators/symm_mem.py (hunk 8 above)
Comment (yuan-luo) [reply python/sglang/srt/distributed/device_communicators/symm_mem.py:95]: The buffer will be auto release when it is not referenced.
https://dev-discuss.pytorch.org/t/pytorch-symmetricmemory-harnessing-nvlink-programmability-with-ease/2798

Comment (yuan-luo) [reply python/sglang/srt/distributed/device_communicators/symm_mem.py:95]: I updated some error branch to deference the buffer.
```
        self.buffer = torch_symm_mem.empty(
            self.max_size // self.dtype.itemsize,
            device=self.device,
            dtype=self.dtype,
        )
        handle = torch_symm_mem.rendezvous(self.buffer, self.group.group_name)
        if handle.multicast_ptr == 0:
            logger.warning(
                "SymmMemCommunicator: symmetric memory "
                "multicast operations are not supported."
            )
            self.buffer = None
            self.disabled = True
            return
        self.disabled = False
```

File: python/sglang/srt/distributed/device_communicators/symm_mem.py (hunk 2 above)
Comment (yuan-luo) [reply python/sglang/srt/distributed/device_communicators/symm_mem.py:51]: Added.

File: python/sglang/srt/distributed/device_communicators/all_reduce_utils.py (hunk 3 above)
Comment (yuan-luo) [reply python/sglang/srt/distributed/device_communicators/all_reduce_utils.py]: Updated to 64MB for tp2 in Blackwell.

File: python/sglang/srt/distributed/device_communicators/all_reduce_utils.py (hunk 7 above)
Comment (yuan-luo) [reply python/sglang/srt/distributed/device_communicators/all_reduce_utils.py:8]: Removed tp16 in test code.

File: python/sglang/srt/distributed/device_communicators/symm_mem.py (hunk 1 above)
Comment (yuan-luo) [reply python/sglang/srt/distributed/device_communicators/symm_mem.py:73]: Removed float32 and float16

File: test/srt/test_symm_mem_allreduce.py (hunk 9 above)
Comment (yuan-luo) [reply test/srt/test_symm_mem_allreduce.py]: Added to ci.

File: test/srt/test_symm_mem_allreduce.py (hunk 10 above)
Comment (yuan-luo) [reply test/srt/test_symm_mem_allreduce.py]: Removed unsupported world sizes and unsupported dtypes.
Currently SymmMemCommunicator will limit the self.max_size to pre-configured max_size. So there's no need to check the exceeding max_size. The upper layer will cover the fallback NCCL logic, but it's not in SymmMemCommunicator's scope.

**Comment by BBuf** [comment]
Resolve conflict.

**Review by BBuf — APPROVED** [review]
LGTM now.

File: python/pyproject.toml [hunk 12]
```diff
@@ -49,6 +49,7 @@ dependencies = [
   "pynvml",
   "python-multipart",
   "pyzmq>=25.1.2",
+  "ray",
```
Comment (merrymercy) [inline python/pyproject.toml]: This dependency is not allowed

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/pyproject.toml (hunk 12 above)
Comment (yuan-luo) [reply python/pyproject.toml]: Removed ray.

File: python/sglang/srt/distributed/device_communicators/symm_mem.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    If any prerequisite is not met, the instance remains disabled and will
+    decline to perform symmetric-memory all-reduce.
+    """
+
+    # Mapping: compute capability major -> supported world sizes for multimem
+    # If the current (cc_major, world_size) is not listed, we fall back
+    # to the two-shot path.
+    _WORLD_SIZES_MULTIMEM = {
+        9: [4, 6, 8],
+        10: [6, 8],
```
Comment (hebiao064) [inline python/sglang/srt/distributed/device_communicators/symm_mem.py:53]: Hi @yuan-luo  I wonder why blackwell doesn't support 4?

**Comment by nvcastet** [comment]
> @trevor-m I knew this PR. Both are two different mechnisms. One uses NCCL and the other uses Torch mechanism.

@yuan-luo 
NCCL production symmetric kernels should be a superset of the custom kernels in PyTorch repo and be more flexible.
PyTorch added those kernels before NCCL support of symmetric memory.
Have you benchmarked your PR against `--enable-symm-mem`? I expect that you will get better perf and also avoid buffer copies.

**Comment by yuan-luo** [comment]
> > @trevor-m I knew this PR. Both are two different mechnisms. One uses NCCL and the other uses Torch mechanism.
> 
> @yuan-luo NCCL production symmetric kernels should be a superset of the custom kernels in PyTorch repo and be more flexible. PyTorch added those kernels before NCCL support of symmetric memory. Have you benchmarked your PR against `--enable-symm-mem`? I expect that you will get better perf and also avoid buffer copies.

@nvcastet Thanks for reminding. Sorry, I've been busy with some stuffs recently. I'll benchmark against `--enable-symm-mem` when available and update here.

File: python/sglang/srt/distributed/device_communicators/symm_mem.py (hunk 13 above)
Comment (yuan-luo) [reply python/sglang/srt/distributed/device_communicators/symm_mem.py:53]: @hebiao064 Blackwell added support 4. https://github.com/sgl-project/sglang/pull/11457
```
    10: {
        2: 64 * MiB,  # 64 MB
        4: 64 * MiB,  # 64 MB
        6: 128 * MiB,  # 128 MB
        8: 128 * MiB,  # 128 MB
    },
```

But if not in the _WORLD_SIZES_MULTIMEM, it will fallback to torch symm mem two shot all reduce.
```
        if self.world_size in self._WORLD_SIZES_MULTIMEM[self.device_capability]:
            torch.ops.symm_mem.multimem_all_reduce_(
                self.buffer[: inp.numel()], "sum", self.group.group_name
            )
        else:
            torch.ops.symm_mem.two_shot_all_reduce_(
                self.buffer[: inp.numel()], "sum", self.group.group_name
            )
```
